### PR TITLE
[SPARK-26721][ML] Avoid per-tree normalization in featureImportance for GBT

### PR DESCRIPTION
**Pull request by mgaido91**
## What changes were proposed in this pull request?

Our feature importance calculation is taken from sklearn's one, which has been recently fixed (in https://github.com/scikit-learn/scikit-learn/pull/11176). Citing the description of that PR:

> Because the feature importances are (currently, by default) normalized and then averaged, feature importances from later stages are overweighted.

The PR performs a fix similar to sklearn's one. The per-tree normalization of the feature importance is skipped and GBT.

Credits for pointing out clearly the issue and the sklearn's PR to Daniel Jumper.

## How was this patch tested?

modified UT, checked that the computed `featureImportance` in that test is similar to sklearn's one (ti can't be the same, because the trees may be slightly different)

### CODE REVIEW

File: mllib/src/main/scala/org/apache/spark/ml/classification/GBTClassifier.scala [hunk 1]
```diff
@@ -341,11 +341,12 @@ class GBTClassificationModel private[ml](
    * The importance vector is normalized to sum to 1. This method is suggested by Hastie et al.
```
Comment (viirya) [inline mllib/src/main/scala/org/apache/spark/ml/classification/GBTClassifier.scala]: This comment is needed to update.

Comment (mgaido91) [reply mllib/src/main/scala/org/apache/spark/ml/classification/GBTClassifier.scala]: no, it is still valid. The final vector is still normalized to 1.

Comment (viirya) [reply mllib/src/main/scala/org/apache/spark/ml/classification/GBTClassifier.scala]: don't you skip normalization of importance vector?

Comment (viirya) [reply mllib/src/main/scala/org/apache/spark/ml/classification/GBTClassifier.scala]: oh, I see. The normalization mentioned here is for total importance.

File: mllib/src/main/scala/org/apache/spark/ml/tree/treeModels.scala [hunk 2]
```diff
@@ -147,18 +147,30 @@ private[ml] object TreeEnsembleModel {
    *                     If -1, then numFeatures is set based on the max feature index in all trees.
    * @return  Feature importance values, of length numFeatures.
    */
-  def featureImportances[M <: DecisionTreeModel](trees: Array[M], numFeatures: Int): Vector = {
```
Comment (viirya) [inline mllib/src/main/scala/org/apache/spark/ml/tree/treeModels.scala]: We need to update the document of `featureImportances`.

File: mllib/src/main/scala/org/apache/spark/ml/classification/GBTClassifier.scala (hunk 1 above)
Comment (mgaido91) [reply mllib/src/main/scala/org/apache/spark/ml/classification/GBTClassifier.scala]: the normalization of the importance vector for each tree, but then at the end the vector is still normalized. To simplify in a diagram, before the PR it was:
`tree importance` -> `normalization` -> `sum` -> `normalization`
now it is
`tree importance` -> `sum` -> `normalization`
So the final result is still normalized.

**Comment by mgaido91** [comment]
@viirya any more comments? cc @srowen

**Review by srowen — APPROVED** [review]

File: mllib/src/test/scala/org/apache/spark/ml/classification/GBTClassifierSuite.scala [hunk 3]
```diff
@@ -363,7 +363,8 @@ class GBTClassifierSuite extends MLTest with DefaultReadWriteTest {
     val gbtWithFeatureSubset = gbt.setFeatureSubsetStrategy("1")
     val importanceFeatures = gbtWithFeatureSubset.fit(df).featureImportances
     val mostIF = importanceFeatures.argmax
-    assert(mostImportantFeature !== mostIF)
+    assert(mostIF === 1)
```
Comment (viirya) [inline mllib/src/test/scala/org/apache/spark/ml/classification/GBTClassifierSuite.scala]: Previously two most important features are different. Why now they are both `1`?

Comment (mgaido91) [reply mllib/src/test/scala/org/apache/spark/ml/classification/GBTClassifierSuite.scala]: Not sure about the exact reason why they were different earlier (of course the behavior changed because of the fix, but this is expected). You can compare the importances vector with the one returned by `sklearn`: as I mentioned in the PR description they are very similar (so `sklearn` too says 1 is the most important in both scenarios using sklearn too).

Comment (srowen) [reply mllib/src/test/scala/org/apache/spark/ml/classification/GBTClassifierSuite.scala]: Yeah I should have commented on this; I actually don't know why the test previously asserted the answers must be different. That's actually the thing I'd least expect, though it's possible. Why does it still assert the importances are different? I suspect they won't match exactly, sure, but if there's an assertion here, isn't it that they're close? They may just not be that comparable in which case there's nothing to assert.

Comment (mgaido91) [reply mllib/src/test/scala/org/apache/spark/ml/classification/GBTClassifierSuite.scala]: The assertion is there to check that a different subset strategy actually produces different results. In particular, in the first case, the importances vector is [1.0, 0.0, ...] while in the second case more features are used (because the trees can check a random variable at time), so the vector is something like [0.7, ...]. Hence this assertion makes sense in order to check that the featureSubset strategy is properly taken in account.

Comment (srowen) [reply mllib/src/test/scala/org/apache/spark/ml/classification/GBTClassifierSuite.scala]: OK, I get it, we just expect something different to happen under the hood, even if we're largely expecting a similar or the same answer. Leave it in; if it failed because it exactly matched, we'd know it, and could easily figure out whether that's actually now expected or a bug.

Comment (viirya) [reply mllib/src/test/scala/org/apache/spark/ml/classification/GBTClassifierSuite.scala]: > In particular, in the first case, the importances vector is [1.0, 0.0, ...] while in the second case more features are used (because the trees can check a random variable at time), so the vector is something like [0.7, ...].

Don't the second case use just one feature and the first case use all features? What you mean more features are used for the second case? Or I misread the test code?

Comment (mgaido91) [reply mllib/src/test/scala/org/apache/spark/ml/classification/GBTClassifierSuite.scala]: In the first case, every tree can choose among all features. Since feature 1 basically is the correct "label" (I mean they are the same), in the first case all the trees choose feature 1 in the first node and they get 100% accuracy. Hence the importance vector is [1.0, 0.0, ...]. In the second case, only 1 random feature per time can be considered. So the trees are more "diverse" and they consider also other features. So the importance vector is the one I mentioned above. You can maybe try and debug this UT if you want to understand better (probably it is more effective than my poor english) or you can try and run the same in sklearn.

Comment (viirya) [reply mllib/src/test/scala/org/apache/spark/ml/classification/GBTClassifierSuite.scala]: Thanks @mgaido91.

I don't have a workable laptop in recent days. So it is hardly for me to run the unit test. That is why I ask for more details.

Sounds that this assertion `assert(importances(mostImportantFeature) !== importanceFeatures(mostIF))` makes sense. But for `assert(mostIF === 1)`, because it picks random feature per time, are we sure that the most importance feature is `1` at all cases? In extreme case, this feature might not be chosen at all. This assertion doesn't make too much sense to me, maybe we don't need it.

Comment (mgaido91) [reply mllib/src/test/scala/org/apache/spark/ml/classification/GBTClassifierSuite.scala]: Well, the seed is fixed, so the UT is actually deterministic and there is no flakyness. Despite with a different seed the result may be different, I'd consider very unlikely anyway that 1 would not be the most important one in any case, since it is really the ground truth in this case.

Comment (viirya) [reply mllib/src/test/scala/org/apache/spark/ml/classification/GBTClassifierSuite.scala]: Yea, it is correct since there is fixed seed.

Anyway, `assert(mostIF === 1)` actually means `assert(mostImportantFeature == mostIF)`. This assertion doesn't make much sense as the previous one `assert(mostImportantFeature !== mostIF)`. It doesn't tell us much except that it is happened to have the same most important feature...

OK for me to leave it as it.

**Review by viirya — APPROVED** [review]

**Review by srowen — APPROVED** [review]

**Comment by srowen** [comment]
Merged to master
